Added FlagSet utility class

peci1 opened the pull request:
This class helps managing a set of flags that are defined as a C++11 enum class.

Adapted from https://github.com/mrts/flag-set-cpp (MIT license, compatible with Apache 2.0).

I have a use for this class in my next PR (not yet published as it needs polishing). But I think it might be generally useful to have a canonical way to work with flag sets.

mjcarroll commented:
This is a bit difficult to evaluate without more context, do you think you could open a draft PR showing the intended use?

peci1 commented:
Okay, here's the example usage:

https://github.com/ignitionrobotics/ign-common/blob/2367c1cb1ae78b4b517efbdbdfa85bc6d525c10f/av/include/ignition/common/HWVideo.hh#L23-L38

https://github.com/ignitionrobotics/ign-common/blob/2367c1cb1ae78b4b517efbdbdfa85bc6d525c10f/av/src/VideoEncoder.cc#L165-L192

mjcarroll commented on include/ignition/common/FlagSet.hh:
```diff
@@ -0,0 +1,348 @@
+  }
+
+  /// \brief Retrurn a FlagSet with all flags set to true.
+  public: static FlagSet ALL()
```
I think also that `All` vs `ALL` may be a bit confusing.

Style Nit: We haven't been using all-caps for static methods (that I'm aware of).

chapulina replied:
No, static methods follow the same `CamelCase` rule as other member functions.

peci1 replied:
Okay, what could be a better name if we'd want to reserve `All()` to the corresponding `all()` function of the bitset? Also `NONE` should be renamed. Some suggestions (I don't have a favorite one yet):

1. AllTrue + AllFalse
1. AllTrue + NoneTrue
1. AllEnabled + NoneEnabled
1. AllEnabled + AllDisabled
1. AllSet + NoneSet

but all of these names still don't clearly differ from the names usually used for getters. The idea was to provide something like `ignition::math::Vector3d::One` to avoid construction patterns like `FlagSet<>().Set()`.

The `NONE` helper is not really needed as the default behavior of the constructor is to set no flags, but I prefered parity of the offered functions.

Another option could be to add a `FlagSet<>(bool)` constructor which would take a parameter telling whether all flags should be set or not. But I think it's less readable and verbose.

mjcarroll replied:
I like `All`, `Any`, `None` as our three checks, as it mirrors STL's bitset, so I think we're good there.

I think my preference for the `static` methods would be `AllEnabled`, `AllDisabled`, but I think it would be worth asking some non-native English speakers to see if it makes as much sense to them.

peci1 replied:
I am non-native :-D But it's a good idea. Where to ask? I assume few people will now have others around to ask them on a coffee break...

mjcarroll replied:
I'm pinging our company chat to recruit some eyeballs.

chapulina replied:
Keeping `All` / `Any` / `Set` as parallels to the `std::bitset` API also make sense to me.

For the static methods, `AllSet` / `NoneSet` make slightly more sense to me - the class is called `FlagSet` afterall. But it's not a strong preference.

peci1 replied:
Okay, renamed according to Louise.

mjcarroll reviewed (commented):
One piece of API feedback in comments.

Does it also make sense to implement `std::hash` the same way that `bitset` does so that it is usable in things like `unordered_map` and `unordered_set`?

peci1 commented:
Do you have a use-case for indexing by bitsets? I can imagine indexing by enum values, but by whole bitsets?

mjcarroll commented:
> Do you have a use-case for indexing by bitsets? I can imagine indexing by enum values, but by whole bitsets?

I was mainly thinking of parity with the `bitset` API, but off hand, here are two things I could think of:

Dispatching based on a certain configuration of flags, so maybe:

```
std::unorderd_map<FlagSet, std::function<void(void)>> dispatch = {
  {FlagSet::AllEnabled, doFoo},
  {FlagSet::AllDisabled, doBar},
};
```

Or potentially keeping  a set of valid configurations to check against:

```
std::set<FlagSet> valid;
valid.contains(FlagSet::AllEnabled);
```

mjcarroll commented:
@peci1 friendly ping

peci1 commented:
I'll have a look at it over Xmas. This week's my vacation.

chapulina commented:
@peci1 , do you need this to go into `ign-common3` so it's available to released collections (Citadel and Dome)? If not, we're starting a new package to hold this kind of utility that doesn't bring any dependencies: https://github.com/ignitionrobotics/ign-utils/. That will be released into Edifice, but it's not clear yet what downstream libraries will depend on it.

chapulina commented:
> do you need this to go into ign-common3

Oh I just saw that you're using it for the video encoder in #125 , so that answers my question.

peci1 commented:
Yes, you're right, I'd like to use this PR in the hardware support for video encoder. But it could be possible to factor this out to a new package starting with Edifice.

peci1 commented:
I did what was requested (renamed the static constructors and added support for std::hash).

I got one more idea - now the class requires that the last element of the enum is an underscore. That is because of two things:

1. determine the size of the bitset in `private: std::bitset<static_cast<UnderlyingType>(T::_)> flags;`
2. disable `FlagSet operator|(const T& _lhs, const T& _rhs)` for enums (or other types) not containing the underscore (as a safety measure, otherwise all types would be OReable to a FlagSet)

I think 1. can be overcome by exposing an explicit template argument for the number of elements in the enum. That would allow much broader use of FlagSet even with enums you do not control. You would just have to pass the correct number of elements. I know this can be difficult to maintain, but many projects use something other than underscore, e.g. `END` or other values to signal the last enum element.

It seems that it is not possible to overcome 2., so let's say only underscore-ending enums would get this syntactic sugar (ORing two base values creates the FlagSet containing both of them). But that's just syntactic sugar.

What do you think about this?

mjcarroll commented:
I don't think I mind having the underscore as the last entry in this case.  That seems easier (maintenance-wise) than having to keep track of the number of elements.  I do see the advantage of exposing the other alternative in the case that it is an enum that we do not maintain/control.

I also agree that disabling the `|` for things without the underscore is reasonable, and that functionality is reserved for enums with the correct sentinel value.

mjcarroll commented:
Friendly ping @peci1, is there anything else you want to do here?

peci1 commented:
I implemented the option for explicit specification of the last element (and whether to treat it as a valid member of the enum or not). I also added some more convenience methods like initializer-list constructor.

I'm satisfied with this PR now (but feel free to comment on the new changes).
